src/ruby/puppetserver-lib/puppet/server/master.rb

```diff
@@ -127,6 +164,79 @@ def getTaskData(environment_name, module_name, task_name)
 
   private
 
+  # @return Puppet::Node::Facts facts, Hash trusted_facts
+  def process_facts(request_data)
+    if request_data['facts'].nil?
+      facts = get_facts_from_pdb(request_data['certname'], request_data['environment'])
+    else
+      facts_from_request = request_data['facts']
+      facts_from_request['name'] = request_data['certname']
+      facts = Puppet::Node::Facts.from_data_hash(facts_from_request)
+    end
+
+    facts.sanitize if facts
+
+    # Pull the trusted facts from the request, or attempt to extract them from
+    # the facts hash
+    trusted_facts = if request_data['trusted_facts']
+                      request_data['trusted_facts']['values']
+                    else
+                      fact_values = facts ? facts.to_data_hash : {}
+                      fact_values['trusted'].nil? ? {} : fact_values['trusted']
+                    end
+
+    return facts, trusted_facts
+  end
+
+  def get_facts_from_pdb(nodename, environment)
+    if Puppet[:storeconfigs] == true
+      pdb_terminus = Puppet::Node::Facts::Puppetdb.new
+      request = Puppet::Indirector::Request.new(pdb_terminus.class.name,
+                                                :find,
+                                                nodename,
+                                                nil,
+                                                :environment => environment)
+      pdb_terminus.find(request)
+    else
+      # How should this be surfaced? Seems like we could maybe do better than a 500, unless
+      # that's accurate?
```

Comment:
I think that depends on user intent, right?

If they sent a request expecting PDB to fill in facts but PDB isn't configured then its a server configuration error. If they wanted to specify their own facts (because they don't use PDB) but forgot, then it's a 400 - Bad Request.

400 feels... nicer, though from the way we've put default usage so far I think a 500 fits better (we just need to very explicitly document our assumptions when we release this).

Reply:
Yeah these were kind of my thoughts. I think given that we can't really know user intent, I think a 500 makes more general sense. I might make the error message more explicitly state what went wrong and the two options: configure PDB or supply your own facts.